rasterio/_io.pyx

```diff
@@ -1107,8 +1109,8 @@ cdef class DatasetReaderBase(DatasetBase):
             exc_wrap_int(
                 GDALGetRasterStatistics(band, int(approx), 1, &min, &max, &mean, &std)
             )
-        except CPLE_BaseError:
-            raise
+        except CPLE_AppDefinedError as exc:
+            raise StatisticsError("No valid pixels found in sampling.")
```

Comment:
Do you think that it would be helpful to propagate the error like this?

```suggestion
            raise StatisticsError("No valid pixels found in sampling.") from exc
```